master/internal/task/resources_list.go

```diff
@@ -1,44 +1,81 @@
 package task
 
 import (
+	"context"
+
+	"github.com/uptrace/bun"
+
+	"github.com/determined-ai/determined/master/internal/db"
 	"github.com/determined-ai/determined/master/internal/sproto"
 	"github.com/determined-ai/determined/master/pkg/cproto"
 	"github.com/determined-ai/determined/master/pkg/device"
+	"github.com/determined-ai/determined/master/pkg/model"
 )
 
 type (
-	// resourcesWithState is an sproto.Resources, along with its state that is tracked by the
+	// ResourcesWithState is an sproto.Resources, along with its state that is tracked by the
 	// allocation. The state is primarily just the updates that come in about the resources.
-	resourcesWithState struct {
-		sproto.Resources
-		rank   int
-		start  *sproto.ResourcesStarted
-		exit   *sproto.ResourcesStopped
-		daemon bool
+	ResourcesWithState struct {
+		bun.BaseModel `bun:"table:allocation_resources,alias:al_res"`
+
+		sproto.Resources `bun:"-"`
+		Rank             int                      `bun:"rank"`
+		Started          *sproto.ResourcesStarted `bun:"started"`
+		Exited           *sproto.ResourcesStopped `bun:"exited"`
+		Daemon           bool                     `bun:"daemon"`
+		ResourceID       sproto.ResourcesID       `bun:"resource_id,type:text"`   // db only
+		AllocationID     model.AllocationID       `bun:"allocation_id,type:text"` // db only
 
 		// The container state, if we're using a RM that uses containers and it was given to us.
 		// This is a rip in the abstraction, remove eventually. Do not add usages.
 		container *cproto.Container
 	}
 
 	// resourcesList tracks resourcesList with their state.
-	resourcesList map[sproto.ResourcesID]*resourcesWithState
+	resourcesList map[sproto.ResourcesID]*ResourcesWithState
 )
 
-func newResourcesState(r sproto.Resources, rank int) resourcesWithState {
-	return resourcesWithState{Resources: r, rank: rank}
+// NewResourcesState creates an instance from `sproto.Resources`.
+func NewResourcesState(r sproto.Resources, rank int) ResourcesWithState {
+	summary := r.Summary()
+	return ResourcesWithState{
+		Resources:    r,
+		Rank:         rank,
+		ResourceID:   summary.ResourcesID,
+		AllocationID: summary.AllocationID,
+	}
+}
+
+// WipeResourcesState deletes all database contents.
+func WipeResourcesState() error {
+	// Bun requires at least one WHERE for updates and deletes.
+	_, err := db.Bun().NewDelete().Model((*ResourcesWithState)(nil)).Where("1=1").Exec(context.TODO())
```

Comment:
could truncate instead?

Reply:
I mean, I don't think we do this anywhere, either. Truncate to me more clearly says 'we're wiping this entire table'. Plus I think (?) it is faster because it works a bit differently internally.

Reply:
truncate is not MVCC-safe, but none of these cleanups run in a tx at the moment.